### PR TITLE
CSPL-1977-CRD support for Readiness, Startup and Liveness Probes

### PR DESCRIPTION
**Pull request by pdhanoyasplunk**
- Added support to configure liveness and readiness probes
- Removed additional delay for app framework phase 2 ES/large apps
- Changed readiness probe to use `/sbin/checkstate.sh` script

### CODE REVIEW

File: pkg/splunk/enterprise/configuration.go [hunk 1]
```diff
@@ -38,6 +38,32 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
+var defaultLivenessProbe corev1.Probe = corev1.Probe{
```
Comment (vivekr-splunk) [inline pkg/splunk/enterprise/configuration.go:41]: effective golang , use short assignment statement `defaultLivenessProbe := corev1.Probe{`

File: pkg/splunk/enterprise/configuration.go [hunk 2]
```diff
@@ -38,6 +38,32 @@ import (
...
+	ProbeHandler: corev1.ProbeHandler{
+		Exec: &corev1.ExecAction{
+			Command: []string{
+				"/sbin/checkstate.sh",
+			},
+		},
+	},
+}
+
+var defaultReadinessProbe corev1.Probe = corev1.Probe{
```
Comment (vivekr-splunk) [inline pkg/splunk/enterprise/configuration.go:54]: effective golang , use short assignment statement `defaultReadinessProbe := corev1.Probe{`

Comment (pdhanoyasplunk) [reply pkg/splunk/enterprise/configuration.go:54]: @vivekr-splunk this is the correct declaration outside of functions. Short declaration is only allowed inside the functions. GoLang will throw an error if I use `:=` here.

**Review by sgontla — APPROVED** [review]

File: pkg/splunk/enterprise/configuration.go (hunk 2 above)
Comment (vivekr-splunk) [reply pkg/splunk/enterprise/configuration.go:54]: oh didnt see they are global

**Review by vivekr-splunk — APPROVED** [review]
LGTM

File: pkg/splunk/enterprise/configuration.go [hunk 3]
```diff
@@ -829,51 +863,54 @@ func updateSplunkPodTemplateWithConfig(ctx context.Context, client splcommon.Con
...
+	readinessProbe := getProbeWithConfigUpdates(&defaultReadinessProbe, spec.ReadinessProbe, spec.ReadinessInitialDelaySeconds)
+	scopedLog.Info("ReadinessProbe", "Configured", readinessProbe)
+	return readinessProbe
+}
 
-	readinessCommand := []string{
-		"/bin/grep",
-		"started",
-		"/opt/container_artifact/splunk-container.state",
+func getProbeWithConfigUpdates(defaultProbe *corev1.Probe, configuredProbe *corev1.Probe, configuredDelay int32) *corev1.Probe {
```
Comment (akondur) [inline pkg/splunk/enterprise/configuration.go:884]: Nit: A small comment on the purpose of the function would be great.

File: pkg/splunk/enterprise/configuration.go [hunk 4]
```diff
@@ -829,51 +863,54 @@ func updateSplunkPodTemplateWithConfig(ctx context.Context, client splcommon.Con
...
+		if derivedProbe.PeriodSeconds == 0 {
+			derivedProbe.PeriodSeconds = defaultProbe.PeriodSeconds
+		}
+		if derivedProbe.Exec == nil {
+			derivedProbe.Exec = defaultProbe.Exec
+		}
+		return &derivedProbe
+	} else if configuredProbe == nil && configuredDelay != 0 {
+		var derivedProbe = *defaultProbe
+		derivedProbe.InitialDelaySeconds = configuredDelay
```
Comment (akondur) [inline pkg/splunk/enterprise/configuration.go:910]: Should we have the `derviedProbe.InitialDelaySeconds = configuredDelay` regardless of configuredProbe to duplicate the assignment ? Seems like that's what we are trying to achieve here?

File: pkg/splunk/enterprise/configuration.go [hunk 5]
```diff
@@ -1607,3 +1644,71 @@ maxGlobalRawDataSizeMB = %d`, indexDefaults, defaults.MaxGlobalRawDataSizeMB)
 `, indexDefaults)
 	return indexDefaults
 }
+
+// validateProbe validates a generic probe values
+func validateProbe(probe *corev1.Probe) error {
```
Comment (akondur) [inline pkg/splunk/enterprise/configuration.go:1649]: Is there an upper threshold for these values?

File: pkg/splunk/enterprise/configuration.go [hunk 6]
```diff
@@ -1607,3 +1644,71 @@ maxGlobalRawDataSizeMB = %d`, indexDefaults, defaults.MaxGlobalRawDataSizeMB)
...
+
+	reqLogger := log.FromContext(ctx)
+	scopedLog := reqLogger.WithName("validatesLivenessProbe").WithValues("name", cr.GetName(), "namespace", cr.GetNamespace())
+
+	err = validateProbe(livenessProbe)
+	if err != nil {
+		return fmt.Errorf("Invalid Liveness Probe config. Reason: %s", err)
+	}
+
+	if livenessProbe.InitialDelaySeconds < livenessProbeDefaultDelaySec {
```
Comment (akondur) [inline pkg/splunk/enterprise/configuration.go:1672]: Can't we incorporate all of this logic inside validateProbe by passing in these parameters? Seems like defaultDelaySec, TimeoutSec, PeriodSec are common to ready and liveness probes?

Comment (pdhanoyasplunk) [reply pkg/splunk/enterprise/configuration.go:1672]: The default values we are comparing against are different. I will explore in future on how to refactor and combine these 2 functions.

File: pkg/splunk/enterprise/configuration.go (hunk 5 above)
Comment (pdhanoyasplunk) [reply pkg/splunk/enterprise/configuration.go:1649]: We dont have any explicit upper limits.

File: pkg/splunk/enterprise/configuration.go [hunk 7]
```diff
@@ -1607,3 +1644,71 @@ maxGlobalRawDataSizeMB = %d`, indexDefaults, defaults.MaxGlobalRawDataSizeMB)
...
+	}
+
+	if livenessProbe.TimeoutSeconds < livenessProbeTimeoutSec {
+		scopedLog.Error(err, "TimeoutSeconds is too small", "configured", livenessProbe.TimeoutSeconds, "recommended minimum", livenessProbeTimeoutSec)
+	}
+
+	if livenessProbe.PeriodSeconds < livenessProbePeriodSec {
+		scopedLog.Error(err, "PeriodSeconds is too small", "configured", livenessProbe.PeriodSeconds, "recommended minimum", livenessProbePeriodSec)
+	}
+
```
[the remark this answers is not in the document]
Comment (pdhanoyasplunk) [reply pkg/splunk/enterprise/configuration.go:1683]: I will take care of the UT in CSPL-2003.

File: pkg/splunk/enterprise/configuration.go (hunk 5 above)
Comment (akondur) [reply pkg/splunk/enterprise/configuration.go:1648]: Should we set it just to safeguard ourselves?

File: pkg/splunk/enterprise/configuration.go (hunk 6 above)
Comment (akondur) [reply pkg/splunk/enterprise/configuration.go:1671]: I was thinking default values can be passed to the validateProbe function which both probes can use? Next sprint is fine too :)

**Review by akondur — APPROVED** [review]

**Review by kashok-splunk — APPROVED** [review]

File: pkg/splunk/enterprise/configuration.go (hunk 4 above)
Comment (pdhanoyasplunk) [reply pkg/splunk/enterprise/configuration.go:911]: That would ignore the scenario when configuredDelay = 0 however deafult Probe has non zero delay.

File: pkg/splunk/enterprise/configuration.go (hunk 5 above)
Comment (pdhanoyasplunk) [reply pkg/splunk/enterprise/configuration.go:1650]: We can take this up in next sprint.

**Review by gaurav-splunk — APPROVED** [review]

**Pull request merged**